File: LeetCode/Python/leetcode_0904_Fruit_Into_Baskets.py
# LeetCode #904 - Fruit Into Baskets
# Date: 2025-08-04

from collections import defaultdict
from typing import List

class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        n = len(fruits)
        mp = defaultdict(int)
        i = 0
        j = 0
        res = 0

        while j < n:
            mp[fruits[j]] += 1

            if len(mp) <= 2:
                res = max(res, j - i + 1)
            else:
                mp[fruits[i]] -= 1
                if mp[fruits[i]] == 0:
                    del mp[fruits[i]]
                i += 1
            j += 1

        return res

File: LeetCode/C++/leetcode_0904_Fruit_Into_Baskets.cpp
// LeetCode #904 - Fruit Into Baskets
// Date: 2025-08-04

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int n = fruits.size();
        unordered_map<int, int> mp;
        int i = 0, j = 0, res = 0;

        while (j < n) {
            mp[fruits[j]]++;

            if (mp.size() <= 2) {
                res = max(res, j - i + 1);
            } else {
                mp[fruits[i]]--;
                if (mp[fruits[i]] == 0) {
                    mp.erase(fruits[i]);
                }
                i++;
            }
            j++;
        }

        return res;
    }
};
